### Lesson 4 Practice Hands-On

### Stepwise Regression Hands On

#### Part I: Backwards Elimination

#### Load in Data

In [ ]:
IQ <- read.csv("/Users/kenyashiafernandez/Desktop/Data-Science-Program/DATASETS/IQ.csv")
head(IQ)

Test1 Test2 Test3 Test4 Test5 IQ 
1 83    34    65    63    64    106
2 73    19    73    48    82     92
3 54    81    82    65    73    102
4 96    72    91    88    94    121
5 84    53    72    68    82    102
6 86    72    63    79    57    105

#### Get a Baseline

In [ ]:
FitAll = lm(IQ ~ ., data = IQ)
summary(IQ)

     Test1           Test2          Test3           Test4           Test5      
 Min.   :37.00   Min.   :19.0   Min.   :43.00   Min.   :39.00   Min.   :42.00  
 1st Qu.:54.00   1st Qu.:51.0   1st Qu.:64.00   1st Qu.:57.50   1st Qu.:59.50  
 Median :71.00   Median :63.0   Median :72.00   Median :68.00   Median :72.00  
 Mean   :67.93   Mean   :61.4   Mean   :72.33   Mean   :65.53   Mean   :69.93  
 3rd Qu.:82.00   3rd Qu.:74.5   3rd Qu.:82.00   3rd Qu.:73.00   3rd Qu.:82.50  
 Max.   :96.00   Max.   :89.0   Max.   :96.00   Max.   :88.00   Max.   :94.00  
       IQ       
 Min.   : 92.0  
 1st Qu.: 96.5  
 Median :102.0  
 Mean   :104.3  
 3rd Qu.:109.0  
 Max.   :130.0  

#### Backward Elimination

In [ ]:
step(FitAll, direction = 'backward')

Start:  AIC=75.31
IQ ~ Test1 + Test2 + Test3 + Test4 + Test5

        Df Sum of Sq    RSS    AIC
- Test5  1      4.61 1025.8 73.378
- Test3  1     25.85 1047.0 73.685
<none>               1021.2 75.310
- Test1  1    400.56 1421.7 78.274
- Test2  1    410.29 1431.5 78.376
- Test4  1    481.32 1502.5 79.103

Step:  AIC=73.38
IQ ~ Test1 + Test2 + Test3 + Test4

        Df Sum of Sq    RSS    AIC
- Test3  1     21.24 1047.0 71.685
<none>               1025.8 73.378
- Test1  1    403.17 1429.0 76.350
- Test2  1    406.47 1432.3 76.384
- Test4  1    481.26 1507.1 77.148

Step:  AIC=71.69
IQ ~ Test1 + Test2 + Test4

        Df Sum of Sq    RSS    AIC
<none>               1047.0 71.685
- Test2  1    406.25 1453.3 74.603
- Test1  1    415.39 1462.4 74.697
- Test4  1    484.13 1531.2 75.386



Call:
lm(formula = IQ ~ Test1 + Test2 + Test4, data = IQ)

Coefficients:
(Intercept)        Test1        Test2        Test4  
     90.733       -1.965       -1.649        3.789  


#### Final Model

In [ ]:
Fitsome <- lm(formula = IQ ~ Test1 + Test2 + Test4, data = IQ)
summary(Fitsome)


Call:
lm(formula = IQ ~ Test1 + Test2 + Test4, data = IQ)

Residuals:
     Min       1Q   Median       3Q      Max 
-10.9184  -6.8179  -0.9142   4.3920  21.1950 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  90.7327    12.8272   7.073 2.06e-05 ***
Test1        -1.9650     0.9406  -2.089   0.0607 .  
Test2        -1.6485     0.7980  -2.066   0.0632 .  
Test4         3.7890     1.6801   2.255   0.0455 *  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 9.756 on 11 degrees of freedom
Multiple R-squared:  0.3839,	Adjusted R-squared:  0.2158 
F-statistic: 2.284 on 3 and 11 DF,  p-value: 0.1356


#### The model with Test 1,2,and 4 is best becasue itg has the lowest ACI of 71.69.
#### From this model the adjusted R2 value is 0.2158, The adjustment is a modification that is supposed to take into account the number of terms in the model.

### Part II: Compare Stepwise Regression Types

#### Load Data

In [ ]:
stepwiseRegression <- read.csv("/Users/kenyashiafernandez/Desktop/Data-Science-Program/DATASETS/stepwiseRegression.csv")

In [ ]:
head(stepwiseRegression)

Y      X1   X2    X3   X4    X5   X6    X7   X8    X9   X10  X11   X12 
1 443.87 74.4 -20.9 29.9  -9.1 -2.3  -0.3 43.3  -5.9 25.7 72.6 -40.3  2.9
2 134.11 75.8 -16.9 33.1 -65.6  0.4 -29.2 49.6  -1.3 11.7 71.5 -37.3  2.7
3 619.19 52.2 -16.9 30.8 -57.4  0.5  34.4 69.9  -5.1 17.9 73.8 -43.7 -1.6
4 465.01 25.8 -23.5 32.4 -52.5 -1.0   4.0 52.1  -9.5 11.9 72.2 -43.7 -1.3
5 609.92 52.8  -7.6 33.5 -34.3 -0.6  33.0 49.2  -1.8 33.5 74.9 -40.4  0.5
6 275.96 73.2 -10.9 27.6 -56.2 -1.6  -7.0 88.8 -16.6 26.4 74.0 -41.4  2.0

#### Backward elimination method

In [ ]:
FitAll = lm(Y ~ ., data = stepwiseRegression)
summary(stepwiseRegression)

       Y                X1              X2               X3       
 Min.   :-53.64   Min.   :25.80   Min.   :-27.60   Min.   :26.60  
 1st Qu.:298.57   1st Qu.:47.48   1st Qu.:-21.73   1st Qu.:30.40  
 Median :422.19   Median :53.65   Median :-18.55   Median :32.10  
 Mean   :427.61   Mean   :54.75   Mean   :-18.15   Mean   :31.97  
 3rd Qu.:565.85   3rd Qu.:62.52   3rd Qu.:-15.18   3rd Qu.:33.35  
 Max.   :947.45   Max.   :80.90   Max.   : -4.90   Max.   :38.00  
       X4               X5                   X6                X7        
 Min.   :-93.40   Min.   :-2.6000000   Min.   :-69.100   Min.   :-14.40  
 1st Qu.:-65.53   1st Qu.:-0.5250000   1st Qu.:-16.550   1st Qu.: 38.52  
 Median :-51.15   Median : 0.0000000   Median :  2.900   Median : 50.30  
 Mean   :-49.85   Mean   :-0.0007813   Mean   :  3.491   Mean   : 48.52  
 3rd Qu.:-35.02   3rd Qu.: 0.6250000   3rd Qu.: 25.100   3rd Qu.: 64.12  
 Max.   :  3.00   Max.   : 3.1000000   Max.   : 70.400   Max.   : 90.00  
       X8    

In [ ]:
step(FitAll, direction = 'backward' , scope = formula(FitAll))

Start:  AIC=220.74
Y ~ X1 + X2 + X3 + X4 + X5 + X6 + X7 + X8 + X9 + X10 + X11 + 
    X12

       Df Sum of Sq     RSS     AIC
- X1    1         2     588  219.07
- X9    1         2     588  219.11
- X8    1         4     590  219.61
- X3    1         5     591  219.87
- X5    1         5     591  219.88
- X7    1         6     592  220.13
<none>                  586  220.74
- X11   1        13     599  221.53
- X2    1        16     602  222.14
- X10   1     20422   21008  676.88
- X12   1    239478  240064  988.69
- X4    1    411780  412366 1057.94
- X6    1   3197390 3197977 1320.13

Step:  AIC=219.07
Y ~ X2 + X3 + X4 + X5 + X6 + X7 + X8 + X9 + X10 + X11 + X12

       Df Sum of Sq     RSS     AIC
- X9    1         2     590  217.53
- X8    1         4     591  217.88
- X3    1         5     592  218.11
- X5    1         5     593  218.15
- X7    1         6     593  218.32
<none>                  588  219.07
- X11   1        13     600  219.81
- X2    1        15     602  220.21
- 


Call:
lm(formula = Y ~ X2 + X4 + X6 + X10 + X11 + X12, data = stepwiseRegression)

Coefficients:
(Intercept)           X2           X4           X6          X10          X11  
 1410.27177     -0.06975      2.80806      5.98660    -11.97829     -0.13102  
        X12  
  -25.98121  


In [ ]:
Fitsome <- lm(formula = Y ~ X2 + X4 + X6 + X10 + X11 + X12, data = stepwiseRegression)
summary(Fitsome)


Call:
lm(formula = Y ~ X2 + X4 + X6 + X10 + X11 + X12, data = stepwiseRegression)

Residuals:
    Min      1Q  Median      3Q     Max 
-5.3190 -1.7438 -0.0897  1.7094  5.3882 

Coefficients:
              Estimate Std. Error  t value Pr(>|t|)    
(Intercept)  1.410e+03  1.320e+01  106.815   <2e-16 ***
X2          -6.975e-02  4.158e-02   -1.677   0.0961 .  
X4           2.808e+00  9.434e-03  297.644   <2e-16 ***
X6           5.987e+00  7.410e-03  807.924   <2e-16 ***
X10         -1.198e+01  1.742e-01  -68.769   <2e-16 ***
X11         -1.310e-01  8.950e-02   -1.464   0.1458    
X12         -2.598e+01  1.171e-01 -221.937   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 2.241 on 121 degrees of freedom
Multiple R-squared:  0.9999,	Adjusted R-squared:  0.9998 
F-statistic: 1.385e+05 on 6 and 121 DF,  p-value: < 2.2e-16


#### Forward Selection

In [ ]:
fitstart = lm(Y ~ 1, data = stepwiseRegression)
summary(fitstart)


Call:
lm(formula = Y ~ 1, data = stepwiseRegression)

Residuals:
    Min      1Q  Median      3Q     Max 
-481.25 -129.04   -5.43  138.24  519.84 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)   427.61      16.03   26.68   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 181.3 on 127 degrees of freedom


In [ ]:
step(fitstart, direction = 'forward', scope = (formula(FitAll)))

Start:  AIC=1332.26
Y ~ 1

       Df Sum of Sq     RSS    AIC
+ X6    1   3312523  862636 1132.4
+ X12   1    565679 3609480 1315.6
+ X4    1    340596 3834563 1323.4
+ X2    1     88438 4086721 1331.5
+ X5    1     77171 4097988 1331.9
<none>              4175159 1332.3
+ X9    1     39558 4135601 1333.0
+ X7    1     39324 4135835 1333.0
+ X1    1     15448 4159711 1333.8
+ X3    1     10545 4164613 1333.9
+ X10   1     10472 4164687 1333.9
+ X11   1      4176 4170982 1334.1
+ X8    1       741 4174418 1334.2

Step:  AIC=1132.41
Y ~ X6

       Df Sum of Sq    RSS     AIC
+ X4    1    584861 277775  989.36
+ X12   1    392692 469944 1056.67
+ X10   1     29736 832900 1129.92
+ X8    1     14958 847678 1132.17
<none>              862636 1132.41
+ X3    1      9152 853483 1133.05
+ X9    1      5341 857295 1133.62
+ X11   1      5119 857517 1133.65
+ X2    1       904 861732 1134.28
+ X1    1       378 862258 1134.36
+ X5    1        63 862573 1134.40
+ X7    1        36 862600 1134.41



Call:
lm(formula = Y ~ X6 + X4 + X12 + X10 + X2 + X11, data = stepwiseRegression)

Coefficients:
(Intercept)           X6           X4          X12          X10           X2  
 1410.27177      5.98660      2.80806    -25.98121    -11.97829     -0.06975  
        X11  
   -0.13102  


#### Final Model

In [ ]:
fitsome = lm(formula = Y ~ X6 + X4 + X12 + X10 + X2 + X11, data = stepwiseRegression)

summary(fitsome)


Call:
lm(formula = Y ~ X6 + X4 + X12 + X10 + X2 + X11, data = stepwiseRegression)

Residuals:
    Min      1Q  Median      3Q     Max 
-5.3190 -1.7438 -0.0897  1.7094  5.3882 

Coefficients:
              Estimate Std. Error  t value Pr(>|t|)    
(Intercept)  1.410e+03  1.320e+01  106.815   <2e-16 ***
X6           5.987e+00  7.410e-03  807.924   <2e-16 ***
X4           2.808e+00  9.434e-03  297.644   <2e-16 ***
X12         -2.598e+01  1.171e-01 -221.937   <2e-16 ***
X10         -1.198e+01  1.742e-01  -68.769   <2e-16 ***
X2          -6.975e-02  4.158e-02   -1.677   0.0961 .  
X11         -1.310e-01  8.950e-02   -1.464   0.1458    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 2.241 on 121 degrees of freedom
Multiple R-squared:  0.9999,	Adjusted R-squared:  0.9998 
F-statistic: 1.385e+05 on 6 and 121 DF,  p-value: < 2.2e-16


#### Hybrid Stepwise - Forward and Backward Selection

In [ ]:
step(fitstart, direction="both", scope=formula(FitAll))

Start:  AIC=1332.26
Y ~ 1

       Df Sum of Sq     RSS    AIC
+ X6    1   3312523  862636 1132.4
+ X12   1    565679 3609480 1315.6
+ X4    1    340596 3834563 1323.4
+ X2    1     88438 4086721 1331.5
+ X5    1     77171 4097988 1331.9
<none>              4175159 1332.3
+ X9    1     39558 4135601 1333.0
+ X7    1     39324 4135835 1333.0
+ X1    1     15448 4159711 1333.8
+ X3    1     10545 4164613 1333.9
+ X10   1     10472 4164687 1333.9
+ X11   1      4176 4170982 1334.1
+ X8    1       741 4174418 1334.2

Step:  AIC=1132.41
Y ~ X6

       Df Sum of Sq     RSS     AIC
+ X4    1    584861  277775  989.36
+ X12   1    392692  469944 1056.67
+ X10   1     29736  832900 1129.92
+ X8    1     14958  847678 1132.17
<none>               862636 1132.41
+ X3    1      9152  853483 1133.05
+ X9    1      5341  857295 1133.62
+ X11   1      5119  857517 1133.65
+ X2    1       904  861732 1134.28
+ X1    1       378  862258 1134.36
+ X5    1        63  862573 1134.40
+ X7    1        36  86


Call:
lm(formula = Y ~ X6 + X4 + X12 + X10 + X2 + X11, data = stepwiseRegression)

Coefficients:
(Intercept)           X6           X4          X12          X10           X2  
 1410.27177      5.98660      2.80806    -25.98121    -11.97829     -0.06975  
        X11  
   -0.13102  


#### Final Model 

In [ ]:
fitsome1 <- lm(formula = Y ~ X6 + X4 + X12 + X10 + X2 + X11, data = stepwiseRegression)
summary(fitsome1)


Call:
lm(formula = Y ~ X6 + X4 + X12 + X10 + X2 + X11, data = stepwiseRegression)

Residuals:
    Min      1Q  Median      3Q     Max 
-5.3190 -1.7438 -0.0897  1.7094  5.3882 

Coefficients:
              Estimate Std. Error  t value Pr(>|t|)    
(Intercept)  1.410e+03  1.320e+01  106.815   <2e-16 ***
X6           5.987e+00  7.410e-03  807.924   <2e-16 ***
X4           2.808e+00  9.434e-03  297.644   <2e-16 ***
X12         -2.598e+01  1.171e-01 -221.937   <2e-16 ***
X10         -1.198e+01  1.742e-01  -68.769   <2e-16 ***
X2          -6.975e-02  4.158e-02   -1.677   0.0961 .  
X11         -1.310e-01  8.950e-02   -1.464   0.1458    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 2.241 on 121 degrees of freedom
Multiple R-squared:  0.9999,	Adjusted R-squared:  0.9998 
F-statistic: 1.385e+05 on 6 and 121 DF,  p-value: < 2.2e-16


#### All final models have an ACI of 213.38 and an adjusted R2 of 0.9998 